sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSink.java

```diff
@@ -146,21 +159,119 @@ public String getMimeType() {
    */
   protected final WritableByteChannelFactory writableByteChannelFactory;
 
+
   /**
-   * Base filename for final output files.
+   * A naming policy for output files.
    */
-  protected final ValueProvider<String> baseOutputFilename;
+  public abstract static class FilenamePolicy
+      implements org.apache.beam.sdk.transforms.SerializableFunction<
+       FilenamePolicy.Context, String> {
+    /**
+     * Context used for generating a name based on window, pane, shard numer, and num shards.
+     * Window and pane will only be provided if windowed writes have been requested using
+     * {@link Write.Bound#withWindowedWrites}. The policy must produce unique filenames for unique
+     * Context objects.
+     */
+    public static class Context {
+      @Nullable private BoundedWindow window;
+      @Nullable private PaneInfo paneInfo;
+      private int shardNumber;
+      private int numShards;
+
+      public Context(
```

Comment:
The only reason these parameters were made nullable was so that sinks that did not yet support windowed writes (HDFS) could throw an exception if the window was not null.